teps/0010-optional-workspaces.md

```diff
@@ -0,0 +1,354 @@
+
+Tekton Variables should fail gracefully in the event that a workspace is
+not provided. Using a variable like `$(workspaces.foo.path)` should not
+cause the controller to abandon or fail a TaskRun when the optional workspace
+`foo` is omitted by it. Likewise the literal string `"$(workspaces.foo.path)"`
+should not end up being passed to the Task.
+
+When a Tekton deployment includes configuration for
+[Default TaskRun Workspaces](https://github.com/tektoncd/pipeline/blob/3a7a693429c605a67e30efa5b74675e403be79bb/config/config-defaults.yaml#L66-L70)
+a missing optional Workspace does not result in a Default TaskRun Workspace
+being injected in its place. The reason for this is that optional behaviour
+in a Task should not be switched on automatically by the Default TaskRun
+Workspace mechanism.
+
+## Proposal
+
+Allow Workspaces to be declared "optional" in Tasks. An optional Workspace is
+not required for successful execution of the Task but can be provided by a TaskRun
+to configure some optional behaviour, like using a specific kind of credential.
+
+### User Stories
+
+#### Story 1
+
+A Catalog Task will be able to support both Workspaces and PipelineResources
+in the same location. This gives greater flexibility to a Tasks' users because
+it allows them to choose whether to use PipelineResources or a Workspace of files
+while the Task behaviour remains the same.
+
+#### Story 2
+
+A TaskRun can be configured to ignore extraneous data. Example: a Task runs
+tests and prints a report. If an optional Workspace is provided it will also
+generate and write a more verbose report to the Workspace.
+
+#### Story 3
+
+A Task will be able to support credentials provided through Tekton's existing
+creds-init mechanism as well as supporting explicit credentials passed via a
+Workspace. Again this provides flexibility: an organization can decide whether
+it's supporting creds-init or credentials via Workspaces and the Task
+behaviour can remain the same.
+
+#### Story 4
+
+A Task should be able to declare a Workspace for a credential that is optional.
+For example, a git Task may be able to utilize a key to access private
+repositories but may also want to support an unauthenticated mode to access
+public repos.
+
+#### Story 5
+
+A Task can accept some optional seed data to modify the way that it works. For
+example, a linter Task can accept an optional configuration file that changes
+some of the default errors into warnings. This would be useful for a mature
+project that has a large body of legacy code that has not been updated with
+latest code styles.
+
+#### Story 6
+
+A Task can accept an optional cache to speed up builds. For example, a webpack
+Task could receive a Workspace with a pre-fetched `node_modules` directory,
+skippingl the `npm install` portion of the build process if one is provided.
+
+### Risks and Mitigations
+
+- tkn may need to be updated to allow users to optionally provide Workspaces
```

Comment:
`tkn` would also need to be updated to show if the workspace is optional in `describe` command

Reply:
Excellent, thankyou, I've added this.